File: cpp/events.cpp
// events subcommand: emit one NDJSON record per accepted assistant turn.
// Reuses cost-mode's parse/dedup/filter/pricing logic; only aggregation
// differs (per-turn NDJSON output instead of accumulated totals).
// See ../SPEC.md §events for the full contract.

#include "events.hpp"
#include "common.hpp"
#include "discovery.hpp"
#include "json_writer.hpp"
#include "pricing.hpp"
#include "walker_roots.hpp"

#include <algorithm>
#include <atomic>
#include <cctype>
#include <charconv>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <optional>
#include <sstream>
#include <string>
#include <string_view>
#include <thread>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include <simdjson.h>

namespace walker::events {

namespace fs = std::filesystem;
namespace sj = simdjson;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

struct EventRecord {
  double ts;
  double usd;
  std::string model;
  std::string session_id;
  std::string slug;
};

// Pricing (rates_for / cost_for) lives in pricing.hpp and the JSON string
// writer in json_writer.hpp — the same shared definitions cost mode and the
// search subcommand use. Both resolve unqualified here via the enclosing
// walker namespace.

// ---------------------------------------------------------------------------
// Argument parsing
// ---------------------------------------------------------------------------

struct Args {
  uint64_t period_seconds = 0;
  double win_start_unix = 0.0;
  bool win_start_set = false;
  std::optional<double> now_unix;
  std::optional<fs::path> projects_root;
  std::vector<fs::path> extra_projects_roots;
  bool read_config = true;
};

[[noreturn]] static void die(std::string_view message) {
  std::cerr << "walker: events: " << message << "\n";
  std::exit(2);
}

static Args parse_args(const std::vector<std::string> &argv) {
  Args args;
  for (size_t i = 0; i < argv.size(); ++i) {
    const std::string &flag = argv[i];
    auto next = [&]() -> const std::string & {
      if (i + 1 >= argv.size())
        die(flag + " needs a value");
      return argv[++i];
    };
    if (flag == "--period") {
      try {
        args.period_seconds = std::stoull(next());
      } catch (...) {
        die("--period: invalid integer");
      }
    } else if (flag == "--win-start") {
      try {
        args.win_start_unix = std::stod(next());
        args.win_start_set = true;
      } catch (...) {
        die("--win-start: invalid number");
      }
    } else if (flag == "--now") {
      try {
        args.now_unix = std::stod(next());
      } catch (...) {
        die("--now: invalid number");
      }
    } else if (flag == "--projects-root") {
      args.projects_root = fs::path(next());
    } else if (flag == "--extra-projects-root") {
      args.extra_projects_roots.emplace_back(next());
    } else if (flag == "--no-config") {
      args.read_config = false;
    } else if (flag == "--version") {
      std::cout << walker::VERSION << "\n";
      std::exit(0);
    } else {
      die(std::string("unknown flag: ") + flag);
    }
  }
  if (args.period_seconds == 0) {
    die("--period is required");
  }
  return args;
}

// ---------------------------------------------------------------------------
// Discovery
// ---------------------------------------------------------------------------

// Shared fused walk + grouping (see discovery.hpp; replaces the per-mode
// copy that iterated each slug dir twice and reused one error_code).
using walker::discover_groups;
using walker::GroupMap;

// ---------------------------------------------------------------------------
// Per-group walker
// ---------------------------------------------------------------------------

static std::vector<EventRecord>
walk_group_events(const std::vector<fs::path> &paths, const std::string &slug,
                  const std::string &session_id, double cutoff) {
  std::vector<EventRecord> records;
  std::unordered_set<std::string> seen_ids;

  sj::ondemand::parser parser;

  for (const auto &path : paths) {
    sj::padded_string data;
    if (sj::padded_string::load(path.string()).get(data) != sj::SUCCESS)
      continue;

    std::string_view buffer(data);
    size_t pos = 0;
    while (pos < buffer.size()) {
      size_t newline = buffer.find('\n', pos);
      size_t end =
          (newline == std::string_view::npos) ? buffer.size() : newline;
      size_t line_end = end;
      if (line_end > pos && buffer[line_end - 1] == '\r')
        --line_end;
      std::string_view line = buffer.substr(pos, line_end - pos);
      pos = (newline == std::string_view::npos) ? buffer.size() : newline + 1;

      // Skip blank lines
      bool blank = true;
      for (char c : line) {
        if (!std::isspace(static_cast<unsigned char>(c))) {
          blank = false;
          break;
        }
      }
      if (blank)
        continue;

      size_t line_off = static_cast<size_t>(line.data() - buffer.data());
      sj::padded_string_view view(line.data(), line.size(),
                                  buffer.size() - line_off +
                                      sj::SIMDJSON_PADDING);
      sj::ondemand::document doc;
      if (parser.iterate(view).get(doc) != sj::SUCCESS)
        continue;

      sj::ondemand::object root;
      if (doc.get_object().get(root) != sj::SUCCESS)
        continue;

      std::string_view timestamp_view;
      bool has_timestamp = false;
      bool is_assistant = false;
      std::string_view message_id_view;
      bool has_message_id = false;
      std::string model;
      uint64_t input_tokens = 0, output_tokens = 0, cache_read_tokens = 0,
               cache_write_tokens = 0, web_search_requests = 0;
      bool message_seen = false;

      for (auto root_field : root) {
        std::string_view key;
        if (root_field.unescaped_key().get(key) != sj::SUCCESS)
          continue;

        if (key == "timestamp") {
          if (root_field.value().get_string().get(timestamp_view) ==
              sj::SUCCESS) {
            has_timestamp = !timestamp_view.empty();
          }
        } else if (key == "message") {
          sj::ondemand::object msg_obj;
          if (root_field.value().get_object().get(msg_obj) != sj::SUCCESS)
            continue;
          message_seen = true;

          for (auto msg_field : msg_obj) {
            std::string_view msg_key;
            if (msg_field.unescaped_key().get(msg_key) != sj::SUCCESS)
              continue;

            if (msg_key == "role") {
              std::string_view role_view;
              if (msg_field.value().get_string().get(role_view) ==
                  sj::SUCCESS) {
                is_assistant = (role_view == "assistant");
              }
            } else if (msg_key == "id") {
              std::string_view id_view;
              if (msg_field.value().get_string().get(id_view) == sj::SUCCESS) {
                if (!id_view.empty()) {
                  message_id_view = id_view;
                  has_message_id = true;
                }
              }
            } else if (msg_key == "model") {
              std::string_view model_view;
              if (msg_field.value().get_string().get(model_view) ==
                  sj::SUCCESS) {
                model.assign(model_view.data(), model_view.size());
                // Lowercase the model string (matches rust:
                // to_ascii_lowercase())
                for (char &c : model) {
                  c = static_cast<char>(
                      std::tolower(static_cast<unsigned char>(c)));
                }
              }
            } else if (msg_key == "usage") {
              sj::ondemand::object usage_obj;
              if (msg_field.value().get_object().get(usage_obj) != sj::SUCCESS)
                continue;

              for (auto usage_field : usage_obj) {
                std::string_view usage_key;
                if (usage_field.unescaped_key().get(usage_key) != sj::SUCCESS)
                  continue;

                // server_tool_use is a nested object, not a scalar.
                // Descend for web_search_requests before the scalar
                // get_uint64 below (which would skip a non-uint value).
                if (usage_key == "server_tool_use") {
                  sj::ondemand::object stu_obj;
                  if (usage_field.value().get_object().get(stu_obj) !=
                      sj::SUCCESS)
                    continue;
                  for (auto stu_field : stu_obj) {
                    std::string_view stu_key;
                    if (stu_field.unescaped_key().get(stu_key) != sj::SUCCESS)
                      continue;
                    if (stu_key == "web_search_requests")
                      web_search_requests =
                          walker::lenient_count(stu_field.value());
                  }
                  continue;
                }

                uint64_t value = walker::lenient_count(usage_field.value());

                if (usage_key == "input_tokens")
                  input_tokens = value;
                else if (usage_key == "output_tokens")
                  output_tokens = value;
                else if (usage_key == "cache_read_input_tokens")
                  cache_read_tokens = value;
                else if (usage_key == "cache_creation_input_tokens")
                  cache_write_tokens = value;
              }
            }
          }
        }
      }

      // Filter 1: must have message with assistant role
      if (!message_seen || !is_assistant)
        continue;

      // Filter 2: dedup by message.id within this group
      if (has_message_id) {
        std::string mid(message_id_view);
        if (!seen_ids.insert(std::move(mid)).second)
          continue;
      }

      // Filter 3: timestamp must parse
      if (!has_timestamp)
        continue;
      auto ts_opt = walker::parse_iso8601(timestamp_view);
      if (!ts_opt)
        continue;
      double ts = *ts_opt;

      // Filter 4: window predicate — ts >= cutoff (= min(now-period,
      // win_start))
      if (ts < cutoff)
        continue;

      double usd = cost_for(input_tokens, output_tokens, cache_read_tokens,
                            cache_write_tokens, web_search_requests, model,
                            timestamp_view);

      records.push_back(EventRecord{ts, usd, model, session_id, slug});
    }
  }

  return records;
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

// Phase timers for WALKER_PROFILE-driven profiling: prints per-phase wall-clock
// to stderr (discover / walk / sort / emit) so the events hotspot is visible.
// Off unless the env var is set; near-zero overhead in production.
namespace {
struct PhaseTimer {
  bool enabled = std::getenv("WALKER_PROFILE") != nullptr;
  std::chrono::steady_clock::time_point last = std::chrono::steady_clock::now();
  void mark(const char *phase) {
    if (!enabled)
      return;
    auto now = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(now - last).count();
    std::cerr << "walker-profile events " << phase << ": " << ms << " ms\n";
    last = now;
  }
};
} // namespace

int run(const std::vector<std::string> &argv) {
  PhaseTimer timer;
  Args args = parse_args(argv);

  double now_unix = args.now_unix.value_or(walker::current_unix());

  // Effective cutoff = min(now - period, win_start), per SPEC §events.
  double period_cutoff = now_unix - static_cast<double>(args.period_seconds);
  // When --win-start is omitted, default to now - period (simplifies the
  // predicate to ts >= now - period, matching rust behavior).
  double win_start = args.win_start_set ? args.win_start_unix : period_cutoff;
  double cutoff = std::min(period_cutoff, win_start);

  fs::path primary =
      args.projects_root.value_or(walker::default_projects_root());
  std::vector<fs::path> roots = walker::resolve_roots(
      primary, args.extra_projects_roots, args.read_config);

  if (roots.empty()) {
    // Primary root doesn't exist — not a hard error; emit no records.
    return 0;
  }

  GroupMap groups = discover_groups(roots, cutoff);
  timer.mark("discover");

  // Decompose group map into a vector of (slug, session_id, paths) tuples
  // for parallel dispatch.
  struct GroupEntry {
    std::string slug;
    std::string session_id;
    std::vector<fs::path> paths;
  };
  std::vector<GroupEntry> group_list;
  group_list.reserve(groups.size());
  for (auto &[key, paths] : groups) {
    // Key is "slug\0session_id" — split on the null separator.
    size_t sep = key.find('\0');
    std::string slug = (sep != std::string::npos) ? key.substr(0, sep) : key;
    std::string sid = (sep != std::string::npos) ? key.substr(sep + 1) : "";
    group_list.push_back(
        GroupEntry{std::move(slug), std::move(sid), std::move(paths)});
  }

  // Parallel walk — mirrors cost-mode's thread pool pattern from main.cpp.
  size_t num_workers =
      walker::effective_workers(std::thread::hardware_concurrency());

  std::vector<std::vector<EventRecord>> per_thread_records(num_workers);
  std::atomic<size_t> task_index(0);

  auto run_tasks = [&](size_t tid) {
    auto &local = per_thread_records[tid];
    while (true) {
      size_t idx = task_index.fetch_add(1, std::memory_order_relaxed);
      if (idx >= group_list.size())
        break;
      auto &ge = group_list[idx];
      auto recs = walk_group_events(ge.paths, ge.slug, ge.session_id, cutoff);
      local.insert(local.end(), std::make_move_iterator(recs.begin()),
                   std::make_move_iterator(recs.end()));
    }
  };

  std::vector<std::thread> threads;
  size_t bg_threads = (num_workers > 1) ? num_workers - 1 : 0;
  threads.reserve(bg_threads);
  for (size_t i = 0; i < bg_threads; ++i) {
    threads.emplace_back(run_tasks, i + 1);
  }
  run_tasks(0); // main thread participates as worker 0
  for (auto &t : threads)
    t.join();

  // Merge per-thread records
  std::vector<EventRecord> all_records;
  {
    size_t total = 0;
    for (auto &v : per_thread_records)
      total += v.size();
    all_records.reserve(total);
    for (auto &v : per_thread_records) {
      all_records.insert(all_records.end(), std::make_move_iterator(v.begin()),
                         std::make_move_iterator(v.end()));
    }
  }
  timer.mark("walk+merge");

  // Sort for deterministic output: (ts, session_id, model) — matches SPEC
  // §events §Ordering and the conformance harness's sort key.
  std::sort(all_records.begin(), all_records.end(),
            [](const EventRecord &a, const EventRecord &b) {
              if (a.ts != b.ts)
                return a.ts < b.ts;
              if (a.session_id != b.session_id)
                return a.session_id < b.session_id;
              return a.model < b.model;
            });
  timer.mark("sort");

  // Emit NDJSON — one line per record. Field order: ts, usd, model, session_id,
  // slug per SPEC §events mandate. Build the whole payload in one buffer and
  // write it with a single fwrite: per-record std::cout/operator<< writes
  // (re-applying the fixed/precision manipulators each field) dominated
  // events-mode wall time.
  std::string out;
  out.reserve(all_records.size() * 128);
  char num[64];
  auto append_fixed6 = [&](double value) {
    // std::to_chars(fixed, 6) is markedly faster than snprintf("%.6f") and
    // produces the identical fixed-6-decimal form for these magnitudes.
    auto [ptr, ec] = std::to_chars(num, num + sizeof(num), value,
                                   std::chars_format::fixed, 6);
    // ts and usd are always finite and small enough that the fixed-6 form
    // fits the 64-byte buffer (token counts are u64-bounded by the lenient
    // parser), so to_chars cannot fail; on the impossible failure the field
    // is simply left empty rather than carrying a dead snprintf fallback.
    if (ec == std::errc())
      out.append(num, static_cast<size_t>(ptr - num));
  };
  for (const auto &r : all_records) {
    out += "{\"ts\":";
    append_fixed6(r.ts);
    out += ",\"usd\":";
    append_fixed6(r.usd);
    out += ",\"model\":";
    write_json_string(out, r.model);
    out += ",\"session_id\":";
    write_json_string(out, r.session_id);
    out += ",\"slug\":";
    write_json_string(out, r.slug);
    out += "}\n";
  }
  std::fwrite(out.data(), 1, out.size(), stdout);
  timer.mark("emit");

  return 0;
}

} // namespace walker::events

File: zig/src/events.zig
// events subcommand: emit one NDJSON record per accepted assistant turn.
// Reuses cost-mode's parse/dedup/filter/pricing verbatim (the shared helpers
// in main.zig); only aggregation differs — per-turn output instead of
// accumulated totals. Mirrors rust/src/events.rs and the cost-mode walk in
// main.zig. See ../SPEC.md §events for the full contract.

const std = @import("std");
const Allocator = std.mem.Allocator;
const main = @import("main.zig");
const walker_roots = @import("walker_roots.zig");

// ─── Args ────────────────────────────────────────────────────────────────────

const EventsArgs = struct {
    period: u64,
    /// Defaults to `now - period` when not supplied by the caller.
    win_start: f64,
    now_unix: f64,
    projects_root: []const u8,
    extra_roots: [][]const u8,
    read_config: bool,
};

fn parseArgs(alloc: Allocator, raw: [][]const u8) !EventsArgs {
    var period: u64 = 0;
    var win_start_raw: ?f64 = null;
    var now_override: ?f64 = null;
    var projects_root: ?[]const u8 = null;
    var extra_roots: std.ArrayList([]const u8) = .empty;
    var read_config = true;

    var i: usize = 0;
    while (i < raw.len) {
        const flag = raw[i];
        i += 1;
        if (std.mem.eql(u8, flag, "--period")) {
            period = std.fmt.parseInt(u64, main.grab(raw, &i, "--period"), 10) catch main.die("events: --period: invalid");
        } else if (std.mem.eql(u8, flag, "--win-start")) {
            win_start_raw = std.fmt.parseFloat(f64, main.grab(raw, &i, "--win-start")) catch main.die("events: --win-start: invalid");
        } else if (std.mem.eql(u8, flag, "--now")) {
            now_override = std.fmt.parseFloat(f64, main.grab(raw, &i, "--now")) catch main.die("events: --now: invalid");
        } else if (std.mem.eql(u8, flag, "--projects-root")) {
            projects_root = main.grab(raw, &i, "--projects-root");
        } else if (std.mem.eql(u8, flag, "--extra-projects-root")) {
            try extra_roots.append(alloc, main.grab(raw, &i, "--extra-projects-root"));
        } else if (std.mem.eql(u8, flag, "--no-config")) {
            read_config = false;
        } else if (std.mem.eql(u8, flag, "--version")) {
            main.writeStdout(main.VERSION ++ "\n");
            std.process.exit(0);
        } else {
            std.debug.print("walker: events: unknown flag: {s}\n", .{flag});
            std.process.exit(2);
        }
    }

    if (period == 0) main.die("events: --period is required");

    const now: f64 = now_override orelse main.nowUnix();
    // When --win-start is omitted, default to now - period (simplifies the
    // predicate to ts >= now - period, per SPEC §events).
    const win_start = win_start_raw orelse (now - @as(f64, @floatFromInt(period)));

    return EventsArgs{
        .period = period,
        .win_start = win_start,
        .now_unix = now,
        .projects_root = projects_root orelse try main.defaultRoot(alloc),
        .extra_roots = try extra_roots.toOwnedSlice(alloc),
        .read_config = read_config,
    };
}

// ─── Output record ───────────────────────────────────────────────────────────

const EventRecord = struct {
    ts: f64,
    usd: f64,
    model: []const u8, // owned (lowercased), valid for arena lifetime
    session_id: []const u8, // borrowed from the discover key
    slug: []const u8, // borrowed from the discover key
};

fn recordLessThan(_: void, a: EventRecord, b: EventRecord) bool {
    // Sort by (ts, session_id, model), matching SPEC §events §Ordering and
    // the conformance multiset tiebreaker.
    if (a.ts != b.ts) return a.ts < b.ts;
    const sid_cmp = std.mem.order(u8, a.session_id, b.session_id);
    if (sid_cmp != .eq) return sid_cmp == .lt;
    return std.mem.order(u8, a.model, b.model) == .lt;
}

// ─── Per-group walker ────────────────────────────────────────────────────────

/// Walk one (slug, session_id) group and append an EventRecord for every
/// accepted assistant turn. Dedup is per-group via a local seen_ids set,
/// exactly matching cost-mode's walk_group contract in main.zig.
fn walkGroupEvents(
    alloc: Allocator,
    paths: []const []const u8,
    slug: []const u8,
    session_id: []const u8,
    cutoff: f64,
    out: *std.ArrayList(EventRecord),
) void {
    var seen = std.StringHashMap(void).init(alloc);
    defer {
        var it = seen.keyIterator();
        while (it.next()) |k| alloc.free(k.*);
        seen.deinit();
    }

    for (paths) |path| {
        const data = main.readEntireFile(alloc, path) catch continue;
        defer alloc.free(data);

        var iter = std.mem.splitScalar(u8, data, '\n');
        while (iter.next()) |line| {
            processLine(alloc, line, &seen, slug, session_id, cutoff, out);
        }
    }
}

fn processLine(
    alloc: Allocator,
    raw: []const u8,
    seen: *std.StringHashMap(void),
    slug: []const u8,
    session_id: []const u8,
    cutoff: f64,
    out: *std.ArrayList(EventRecord),
) void {
    const line = std.mem.trim(u8, raw, " \t\r\n");
    if (line.len == 0) return;

    var scanner = std.json.Scanner.initCompleteInput(alloc, line);
    defer scanner.deinit();

    if (!(main.enterObject(&scanner) catch return)) return;

    var role_assistant = false;
    var id_str: ?[]const u8 = null;
    var ts_value: ?f64 = null;
    // Raw timestamp string, retained for date-aware sonnet-5 pricing.
    var ts_date: []const u8 = "";
    var model: []const u8 = "";
    var inp: u64 = 0;
    var out_: u64 = 0;
    var cr: u64 = 0;
    var cw: u64 = 0;
    var web_searches: u64 = 0;

    while (true) {
        const key = (main.parseObjectKey(&scanner, alloc) catch return) orelse break;
        if (std.mem.eql(u8, key, "message")) {
            if (!(main.enterObject(&scanner) catch return)) continue;
            while (true) {
                const mkey = (main.parseObjectKey(&scanner, alloc) catch return) orelse break;
                if (std.mem.eql(u8, mkey, "role")) {
                    const v = main.parseStringValue(&scanner, alloc) catch return;
                    if (v) |s| role_assistant = std.mem.eql(u8, s, "assistant");
                } else if (std.mem.eql(u8, mkey, "id")) {
                    id_str = main.parseStringValue(&scanner, alloc) catch return;
                } else if (std.mem.eql(u8, mkey, "model")) {
                    const v = main.parseStringValue(&scanner, alloc) catch return;
                    if (v) |s| model = s;
                } else if (std.mem.eql(u8, mkey, "usage")) {
                    if (!(main.enterObject(&scanner) catch return)) continue;
                    while (true) {
                        const ukey = (main.parseObjectKey(&scanner, alloc) catch return) orelse break;
                        if (std.mem.eql(u8, ukey, "input_tokens")) {
                            inp = main.parseU64Value(&scanner, alloc) catch return;
                        } else if (std.mem.eql(u8, ukey, "output_tokens")) {
                            out_ = main.parseU64Value(&scanner, alloc) catch return;
                        } else if (std.mem.eql(u8, ukey, "cache_read_input_tokens")) {
                            cr = main.parseU64Value(&scanner, alloc) catch return;
                        } else if (std.mem.eql(u8, ukey, "cache_creation_input_tokens")) {
                            cw = main.parseU64Value(&scanner, alloc) catch return;
                        } else if (std.mem.eql(u8, ukey, "server_tool_use")) {
                            // Nested object: descend for web_search_requests.
                            if (main.enterObject(&scanner) catch return) {
                                while (true) {
                                    const skey = (main.parseObjectKey(&scanner, alloc) catch return) orelse break;
                                    if (std.mem.eql(u8, skey, "web_search_requests")) {
                                        web_searches = main.parseU64Value(&scanner, alloc) catch return;
                                    } else {
                                        scanner.skipValue() catch return;
                                    }
                                }
                            }
                        } else {
                            scanner.skipValue() catch return;
                        }
                    }
                } else {
                    scanner.skipValue() catch return;
                }
            }
        } else if (std.mem.eql(u8, key, "timestamp")) {
            const v = main.parseStringValue(&scanner, alloc) catch return;
            if (v) |s| {
                ts_value = main.parseTs(s) catch null;
                ts_date = s;
            }
        } else {
            scanner.skipValue() catch return;
        }
    }

    // Filter 1: assistant role only.
    if (!role_assistant) return;

    // Filter 2: dedup by message.id within this group.
    if (id_str) |id| {
        if (id.len > 0) {
            if (seen.contains(id)) return;
            // alloc is a worker arena: on a put failure the duped key is
            // reclaimed at arena deinit, so no manual free is needed.
            const k = alloc.dupe(u8, id) catch return;
            seen.put(k, {}) catch return;
        }
    }

    // Filter 3 + 4: timestamp must parse and pass the window predicate.
    const ts = ts_value orelse return;
    if (ts < cutoff) return;

    const usd = main.modelCost(inp, out_, cr, cw, web_searches, model, ts_date);

    // model is emitted lowercased (SPEC: "Lowercased model id ...").
    const model_lower = alloc.alloc(u8, model.len) catch return;
    _ = std.ascii.lowerString(model_lower, model);

    out.append(alloc, .{
        .ts = ts,
        .usd = usd,
        .model = model_lower,
        .session_id = session_id,
        .slug = slug,
    }) catch return;
}

// ─── JSON string escaping ────────────────────────────────────────────────────

fn appendJsonString(buf: *std.ArrayList(u8), alloc: Allocator, s: []const u8) !void {
    try buf.append(alloc, '"');
    for (s) |c| {
        switch (c) {
            '"' => try buf.appendSlice(alloc, "\\\""),
            '\\' => try buf.appendSlice(alloc, "\\\\"),
            '\n' => try buf.appendSlice(alloc, "\\n"),
            '\r' => try buf.appendSlice(alloc, "\\r"),
            '\t' => try buf.appendSlice(alloc, "\\t"),
            8 => try buf.appendSlice(alloc, "\\b"),
            12 => try buf.appendSlice(alloc, "\\f"),
            else => {
                if (c < 0x20) {
                    const escaped = try std.fmt.allocPrint(alloc, "\\u{x:0>4}", .{c});
                    try buf.appendSlice(alloc, escaped);
                } else {
                    try buf.append(alloc, c);
                }
            },
        }
    }
    try buf.append(alloc, '"');
}

// ─── Parallel walk infrastructure ────────────────────────────────────────────

// One unit of work: a discovered (slug, session_id) group and its file paths.
const GroupWork = struct {
    slug: []const u8,
    session_id: []const u8,
    paths: []const []const u8,
};

// Lock-free group queue: workers fetchAdd a cursor and take the indexed group.
// Mirrors the cost-mode Queue in main.zig.
const EventsQueue = struct {
    items: []const GroupWork,
    cur: std.atomic.Value(usize),
    fn init(items: []const GroupWork) EventsQueue {
        return .{ .items = items, .cur = .init(0) };
    }
    fn pop(self: *EventsQueue) ?GroupWork {
        const i = self.cur.fetchAdd(1, .seq_cst);
        return if (i < self.items.len) self.items[i] else null;
    }
};

// Per-worker state: a private arena (so threads never share an allocator) and
// the records it collected. The arena outlives the merge+emit below (its
// deinit is deferred in run after emit), so EventRecord.model — allocated here
// — stays valid; session_id/slug are borrowed from the run-level discover keys.
const WorkerSlot = struct {
    arena: std.heap.ArenaAllocator,
    records: std.ArrayList(EventRecord) = .empty,
};

fn worker(queue: *EventsQueue, slot: *WorkerSlot, cutoff: f64) void {
    const alloc = slot.arena.allocator();
    while (queue.pop()) |gw| {
        walkGroupEvents(alloc, gw.paths, gw.slug, gw.session_id, cutoff, &slot.records);
    }
}

// ─── Entry point ─────────────────────────────────────────────────────────────

pub fn run(gpa: Allocator, argv: [][]const u8) !void {
    var arena = std.heap.ArenaAllocator.init(gpa);
    defer arena.deinit();
    const alloc = arena.allocator();

    const args = try parseArgs(alloc, argv);

    // Effective cutoff = min(now - period, win_start), per SPEC §events.
    const period_cutoff = args.now_unix - @as(f64, @floatFromInt(args.period));
    const cutoff = @min(period_cutoff, args.win_start);

    const roots = try walker_roots.resolveRoots(alloc, args.projects_root, args.extra_roots, args.read_config);
    if (roots.len == 0) {
        // Primary root doesn't exist — emit nothing, exit 0 (consistent with
        // cost-mode's empty-fleet case).
        return;
    }

    // discover applies the mtime prune using the same cutoff.
    var grp_map = try main.discover(alloc, roots, cutoff);

    // Flatten the discover map into a work list. Keys are "{slug}\x00{sid}"
    // (see main.addFile); split to recover slug and session_id.
    var work: std.ArrayList(GroupWork) = .empty;
    var it = grp_map.iterator();
    while (it.next()) |kv| {
        const key = kv.key_ptr.*;
        const sep = std.mem.indexOfScalar(u8, key, 0) orelse continue;
        try work.append(alloc, .{
            .slug = key[0..sep],
            .session_id = key[sep + 1 ..],
            .paths = kv.value_ptr.*.items,
        });
    }

    var records: std.ArrayList(EventRecord) = .empty;

    // Parallel walk, capped at 8 workers (matching cost mode). Events was
    // previously a serial loop here, which left it ~9x slower than the other
    // impls on a full fleet — the walk, not the emit, was the bottleneck.
    // Always uses the queue+thread shape (a single worker thread when
    // ncpu==1), matching cost/search/beacons-history; the old serial
    // special-case was a second, divergent code path for the same work.
    const ncpu = std.Thread.getCpuCount() catch 4;
    var nw: usize = @min(@as(usize, 8), @max(@as(usize, 1), ncpu));
    if (nw > work.items.len) nw = work.items.len;
    if (nw == 0) nw = 1; // empty work list: one worker drains immediately

    // The worker arenas own the EventRecord.model strings, so they must outlive
    // the emit below. Declare the cleanup at function scope - a block-scoped
    // defer would free them before emit reads rec.model, a use-after-free.
    var queue = EventsQueue.init(work.items);
    const slots = try alloc.alloc(WorkerSlot, nw);
    for (slots) |*s| s.* = .{ .arena = std.heap.ArenaAllocator.init(std.heap.page_allocator) };
    defer for (slots) |*s| s.arena.deinit();

    var threads: std.ArrayList(std.Thread) = .empty;
    for (slots) |*s| try threads.append(alloc, try std.Thread.spawn(.{}, worker, .{ &queue, s, cutoff }));
    for (threads.items) |t| t.join();

    for (slots) |*s| try records.appendSlice(alloc, s.records.items);

    // Sort for deterministic output: (ts, session_id, model).
    std.mem.sort(EventRecord, records.items, {}, recordLessThan);

    // Emit NDJSON — one line per record, field order fixed per SPEC:
    // ts, usd, model, session_id, slug. Numbers format into a reused stack
    // buffer (no per-field heap alloc) and the whole payload is written once.
    var buf: std.ArrayList(u8) = .empty;
    try buf.ensureTotalCapacity(alloc, records.items.len * 128);
    var num_buf: [64]u8 = undefined;
    for (records.items) |rec| {
        try buf.appendSlice(alloc, "{\"ts\":");
        try buf.appendSlice(alloc, std.fmt.bufPrint(&num_buf, "{d}", .{rec.ts}) catch unreachable);
        try buf.appendSlice(alloc, ",\"usd\":");
        try buf.appendSlice(alloc, std.fmt.bufPrint(&num_buf, "{d}", .{rec.usd}) catch unreachable);
        try buf.appendSlice(alloc, ",\"model\":");
        try appendJsonString(&buf, alloc, rec.model);
        try buf.appendSlice(alloc, ",\"session_id\":");
        try appendJsonString(&buf, alloc, rec.session_id);
        try buf.appendSlice(alloc, ",\"slug\":");
        try appendJsonString(&buf, alloc, rec.slug);
        try buf.appendSlice(alloc, "}\n");
    }
    main.writeStdout(buf.items);
}
